File: electron-app/src/UserManager.ts
import { BrowserWindow, ipcMain } from 'electron';
import { BackendManager } from './BackendManager';
import * as path from 'path';
import * as fs from 'fs';

export class UserManager {
  private backendManager: BackendManager;
  private userDataPath: string;
  private isDev: boolean;

  constructor(backendManager: BackendManager, userDataPath: string, isDev: boolean) {
    this.backendManager = backendManager;
    this.userDataPath = userDataPath;
    this.isDev = isDev;
  }

  async checkAndHandleUsers(pythonPath: string, venvPath: string, parentWindow?: BrowserWindow): Promise<void> {
    try {
      const backendDir = path.join(__dirname, '..', 'backend');
      const userCount = await this.getUserCount(backendDir, venvPath);

      console.log(`[UserManager] Found ${userCount} users in database`);

      if (userCount === 0) {
        console.log('[UserManager] No users found, setting up superuser creation...');
        await this.showSuperuserCreationModal(backendDir, venvPath, parentWindow);
      } else {
        console.log(`[UserManager] Found ${userCount} existing users, no action needed`);
      }
    } catch (error) {
      console.error('[UserManager] Error checking users:', error);
      throw error;
    }
  }

  private async getUserCount(backendDir: string, venvPath: string): Promise<number> {
    const pythonCode = 'from django.contrib.auth.models import User; print(User.objects.count())';
    const output = await this.backendManager.runDjangoShellCommand(backendDir, venvPath, pythonCode);

    // Parse the output to get user count
    const lines = output.trim().split('\n').filter(line => line.trim() !== '');
    const userCountStr = lines[lines.length - 1];
    const userCount = parseInt(userCountStr);

    return isNaN(userCount) ? 0 : userCount;
  }

  private async showSuperuserCreationModal(backendDir: string, venvPath: string, parentWindow?: BrowserWindow): Promise<void> {
    return new Promise<void>((resolve) => {
      let promptWindow: BrowserWindow | null = new BrowserWindow({
        width: 500,
        height: 650,
        modal: true,
        parent: parentWindow,
        show: false,
        resizable: false,
        titleBarStyle: 'hidden',
        titleBarOverlay: process.platform !== 'darwin',
        webPreferences: {
          nodeIntegration: false,
          contextIsolation: true,
          preload: path.join(__dirname, 'superuser-modal-preload.js')
        }
      });

      // Load HTML from file
      const htmlPath = path.join(__dirname, 'superuser-modal.html');
      promptWindow.loadFile(htmlPath);

      promptWindow.once('ready-to-show', () => {
        promptWindow.show();
      });

      // Handle superuser creation
      ipcMain.once('create-superuser', async (event, userData) => {
        try {
          await this.createSuperuser(backendDir, venvPath, userData.username, userData.email, userData.password);
          event.reply('superuser-created', true, 'Superuser created successfully');

          // After successful superuser creation, check for schemas and column templates
          setTimeout(async () => {
            if (promptWindow && !promptWindow.isDestroyed()) {
              promptWindow.close();
            }
            await this.showManagementPanelIfNeeded(backendDir, venvPath, parentWindow);
          }, 1000);
        } catch (error) {
          event.reply('superuser-created', false, error.message);
        }
      });

      // Handle cancel
      ipcMain.once('cancel-superuser', () => {
        if (promptWindow && !promptWindow.isDestroyed()) {
          promptWindow.close();
        }
      });

      // Handle close window
      ipcMain.once('close-superuser-window', () => {
        if (promptWindow && !promptWindow.isDestroyed()) {
          promptWindow.close();
        }
      });

      promptWindow.on('closed', () => {
        promptWindow = null;
        resolve();
      });
    });
  }

  private async createSuperuser(backendDir: string, venvPython: string, username: string, email: string, password: string): Promise<void> {
    const pythonCode = `
from django.contrib.auth.models import User
try:
    user = User.objects.create_superuser('${username}', '${email}', '${password}')
    print(f'Superuser "${username}" created successfully')
except Exception as e:
    print(f'Error creating superuser: {e}')
    exit(1)
`;

    const output = await this.backendManager.runDjangoShellCommand(backendDir, venvPython, pythonCode);

    if (!output.includes('created successfully')) {
      throw new Error(`Failed to create superuser: ${output}`);
    }

    console.log(`[UserManager] Superuser "${username}" created successfully`);
  }

  private async showManagementPanelIfNeeded(backendDir: string, venvPath: string, parentWindow?: BrowserWindow): Promise<void> {
    try {
      const needsSchemas = !(await this.backendManager.checkSchemas(backendDir, venvPath));
      const needsColumnTemplates = !(await this.backendManager.checkColumnTemplates(backendDir, venvPath));

      if (needsSchemas || needsColumnTemplates) {
        await this.showManagementPanel(backendDir, venvPath, needsSchemas, needsColumnTemplates, parentWindow);
      }
    } catch (error) {
      console.error('[UserManager] Error checking management requirements:', error);
    }
  }

  async showManagementPanel(backendDir: string, venvPath: string, needsSchemas: boolean, needsColumnTemplates: boolean, parentWindow?: BrowserWindow): Promise<void> {
    return new Promise<void>((resolve) => {
      let managementWindow: BrowserWindow | null = new BrowserWindow({
        width: 600,
        height: 500,
        modal: true,
        parent: parentWindow,
        show: false,
        resizable: false,
        titleBarStyle: 'hidden',
        titleBarOverlay: process.platform !== 'darwin',
        webPreferences: {
          nodeIntegration: false,
          contextIsolation: true,
          preload: path.join(__dirname, 'management-panel-preload.js')
        }
      });

      const htmlPath = path.join(__dirname, 'management-panel.html');
      managementWindow.loadFile(htmlPath);

      managementWindow.once('ready-to-show', () => {
        managementWindow.show();
        managementWindow.webContents.send('setup-panel', { needsSchemas, needsColumnTemplates });
      });

      // Handle schema initialization
      ipcMain.once('run-sync-schemas', async (event) => {
        try {
          event.reply('command-progress', 'sync-schemas', 'running', 'Running sync_schemas command...');
          await this.backendManager.runManagementCommand(backendDir, venvPath, 'sync_schemas');
          event.reply('command-progress', 'sync-schemas', 'completed', 'Schema synchronization completed successfully');
        } catch (error) {
          event.reply('command-progress', 'sync-schemas', 'error', `Schema synchronization failed: ${error.message}`);
        }
      });

      // Handle column template initialization
      ipcMain.once('run-load-column-templates', async (event) => {
        try {
          event.reply('command-progress', 'load-column-templates', 'running', 'Running load_column_templates command...');
          await this.backendManager.runManagementCommand(backendDir, venvPath, 'load_column_templates');
          event.reply('command-progress', 'load-column-templates', 'completed', 'Column templates loaded successfully');
        } catch (error) {
          event.reply('command-progress', 'load-column-templates', 'error', `Column templates loading failed: ${error.message}`);
        }
      });

      // Handle skip
      ipcMain.once('skip-management', () => {
        if (managementWindow && !managementWindow.isDestroyed()) {
          managementWindow.close();
        }
      });

      // Handle close window
      ipcMain.once('close-management-window', () => {
        if (managementWindow && !managementWindow.isDestroyed()) {
          managementWindow.close();
        }
      });

      managementWindow.on('closed', () => {
        managementWindow = null;
        resolve();
      });
    });
  }
}

File: cupcake-vanilla-wails/backend/services/user_manager.go
package services

import (
	"fmt"
	"log"
	"os/exec"
	"path/filepath"
	"strconv"
	"strings"
)

type UserManager struct {
	backendManager *BackendManager
	userDataPath   string
	isDev          bool
}

func NewUserManager(backendManager *BackendManager, userDataPath string, isDev bool) *UserManager {
	return &UserManager{
		backendManager: backendManager,
		userDataPath:   userDataPath,
		isDev:          isDev,
	}
}

func (u *UserManager) GetUserCount(backendDir, pythonPath string) (int, error) {
	log.Printf("[UserManager] GetUserCount: backendDir=%s, pythonPath=%s", backendDir, pythonPath)

	managePy := filepath.Join(backendDir, "manage.py")
	script := "from django.contrib.auth import get_user_model; User = get_user_model(); print(User.objects.count())"

	cmd := exec.Command(pythonPath, managePy, "shell", "-c", script)
	cmd.Dir = backendDir
	cmd.Env = u.getBackendEnv(backendDir)

	output, err := cmd.CombinedOutput()
	log.Printf("[UserManager] GetUserCount output: %s", string(output))
	if err != nil {
		log.Printf("[UserManager] Error getting user count: %v, output: %s", err, string(output))
		return 0, err
	}

	countStr := strings.TrimSpace(string(output))
	log.Printf("[UserManager] Parsing count from: '%s'", countStr)

	lines := strings.Split(countStr, "\n")
	lastLine := strings.TrimSpace(lines[len(lines)-1])
	log.Printf("[UserManager] Last line: '%s'", lastLine)

	count, err := strconv.Atoi(lastLine)
	if err != nil {
		log.Printf("[UserManager] Failed to parse count from: '%s'", lastLine)
		return 0, fmt.Errorf("failed to parse user count: %s", countStr)
	}

	log.Printf("[UserManager] User count: %d", count)
	return count, nil
}

func (u *UserManager) CreateSuperuser(backendDir, pythonPath, username, email, password string) error {
	log.Printf("[UserManager] Creating superuser: %s", username)
	log.Printf("[UserManager] Backend dir: %s", backendDir)
	log.Printf("[UserManager] Python path: %s", pythonPath)

	managePy := filepath.Join(backendDir, "manage.py")
	log.Printf("[UserManager] manage.py path: %s", managePy)

	script := fmt.Sprintf(`
from django.contrib.auth import get_user_model
User = get_user_model()
if not User.objects.filter(username='%s').exists():
    User.objects.create_superuser('%s', '%s', '%s')
    print('SUCCESS')
else:
    print('EXISTS')
`, username, username, email, password)

	cmd := exec.Command(pythonPath, managePy, "shell", "-c", script)
	cmd.Dir = backendDir
	cmd.Env = u.getBackendEnv(backendDir)

	log.Printf("[UserManager] Executing command: %s %s shell -c ...", pythonPath, managePy)

	output, err := cmd.CombinedOutput()
	log.Printf("[UserManager] Command output: %s", string(output))
	if err != nil {
		log.Printf("[UserManager] Error creating superuser: %v, output: %s", err, string(output))
		return fmt.Errorf("failed to create superuser: %w", err)
	}

	result := strings.TrimSpace(string(output))
	if strings.Contains(result, "SUCCESS") {
		log.Printf("[UserManager] Superuser %s created successfully", username)
		return nil
	}
	if strings.Contains(result, "EXISTS") {
		log.Printf("[UserManager] Superuser %s already exists", username)
		return nil
	}

	log.Printf("[UserManager] Unexpected output: %s", result)
	return nil
}

func (u *UserManager) CheckUserExists(backendDir, pythonPath, username string) (bool, error) {
	managePy := filepath.Join(backendDir, "manage.py")
	script := fmt.Sprintf(`
from django.contrib.auth import get_user_model
User = get_user_model()
print('EXISTS' if User.objects.filter(username='%s').exists() else 'NOT_EXISTS')
`, username)

	cmd := exec.Command(pythonPath, managePy, "shell", "-c", script)
	cmd.Dir = backendDir
	cmd.Env = u.getBackendEnv(backendDir)

	output, err := cmd.CombinedOutput()
	if err != nil {
		return false, err
	}

	return strings.Contains(strings.TrimSpace(string(output)), "EXISTS"), nil
}

func (u *UserManager) ChangePassword(backendDir, pythonPath, username, newPassword string) error {
	managePy := filepath.Join(backendDir, "manage.py")
	script := fmt.Sprintf(`
from django.contrib.auth import get_user_model
User = get_user_model()
user = User.objects.get(username='%s')
user.set_password('%s')
user.save()
print('SUCCESS')
`, username, newPassword)

	cmd := exec.Command(pythonPath, managePy, "shell", "-c", script)
	cmd.Dir = backendDir
	cmd.Env = u.getBackendEnv(backendDir)

	output, err := cmd.CombinedOutput()
	if err != nil {
		log.Printf("[UserManager] Error changing password: %v, output: %s", err, string(output))
		return fmt.Errorf("failed to change password: %w", err)
	}

	return nil
}

func (u *UserManager) ListUsers(backendDir, pythonPath string) ([]string, error) {
	managePy := filepath.Join(backendDir, "manage.py")
	script := `
from django.contrib.auth import get_user_model
User = get_user_model()
for user in User.objects.all():
    print(user.username)
`

	cmd := exec.Command(pythonPath, managePy, "shell", "-c", script)
	cmd.Dir = backendDir
	cmd.Env = u.getBackendEnv(backendDir)

	output, err := cmd.CombinedOutput()
	if err != nil {
		return nil, err
	}

	lines := strings.Split(strings.TrimSpace(string(output)), "\n")
	var users []string
	for _, line := range lines {
		line = strings.TrimSpace(line)
		if line != "" {
			users = append(users, line)
		}
	}

	return users, nil
}

func (u *UserManager) getBackendEnv(backendDir string) []string {
	if u.backendManager != nil {
		return u.backendManager.getBackendEnv(backendDir)
	}
	return nil
}
